File: Numerical Series/Fibonacci.jl
#### Fibonacci Sequence

"""
f(n) =  1                   (n = 1)
        1                   (n = 2)
        f(n-1) + f(n-2)     (n > 2)
"""

## Fibonacci Sequence with Number

using Plots; pyplot()

function fib1(n::Int64)
    a = 1
    b = 1
    c = 1
    if n <= 0
        throw(DomainError(n, "Please input a positive integer."))
        return -1
    end
    while (n-2) > 0
        c = a + b
        a = b
        b = c
        n = n - 1
    end
    return c
end

function fib2(n::Int64)
    a = 1
    b = 1
    c = 1
    if n <= 0
        throw(DomainError(n, "Please input a positive integer."))
        return -1
    end
    for i in 3:n
        c = a + b
        a = b
        b = c
    end
    return c
end

function fib3(n::Int64)
    if n <= 0
        throw(DomainError(n, "Please input a positive integer."))
        return -1
    elseif 1 <= n <= 2
        return 1
    else
        return fib1(n-1) + fib1(n-2)
    end
end

function fib4(n::Int64)
    if n <= 0
        throw(DomainError(n, "Please input a positive integer."))
        return -1
    elseif 1 <= n <= 2
        return 1
    end
    return fib2(n-1) + fib2(n-2)
end

println("Please input a positive integer.")
number = parse(Int64, chomp(readline()))
to = time()

re = fib1(number)

if re != -1
    println("The fibonacci sequence of $number is $re.")
end
td = time() - to
println("The time interval is $td s.")

## Fibonacci Sequence in Array

to = time()

N = 1:1:100
M = [fib1(n) for n in N]

td = time() - to
print("The time interval is $td s.")

plot(fontfamily=("Serif"),dpi=1024)
plot!(N,M,color="#00B4DC",label="Fibonacci Sequence")

File: Numerical Series/Fibonacci.py
#### Fibonacci Sequence

"""
f(n) =  1                   (n = 1)
        1                   (n = 2)
        f(n-1) + f(n-2)     (n > 2)
"""

#%%  Fibonacci With Iteration

import time
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from matplotlib import ticker

number = int(input("Please input a positive integer. \n"))
to = time.time()

def fib(n):
    a = 1
    b = 1
    c = 1
    if n < 1:
        print("Please a positive integer.")
        return -1
    while (n-2) > 0:
        c = b + a
        a = b
        b = c
        n = n - 1
    return c

re = fib(number)
if re != -1:
    print("The fibonacci sequence of %d is %d." %(number,re))
tf = time.time()
td = tf - to
print("Fibonacci With Iteration: The time interval is %f s." %td)

#%%  Fibonacci With Recursion

number = int(input("Please input a positive integer. \n"))
to = time.time()

def fib_r(n):
    if n < 1:
        print("Please a positive integer.")
        return -1
    if n == 1 or n == 2:
        return 1
    else:
        return fib_r(n-1) + fib_r(n-2)

re = fib_r(number)
if re != -1:
    print("The fibonacci sequence of %d is %d." %(number,re))
tf = time.time()
td = tf - to
print("Fibonacci With Recursion: The time interval is %f s." %td)

#%%  Fibonacci Sequence in Array

to = time.time()

upLim = 40
N = np.arange(1,upLim+1,1)
M = np.array(list(map(fib,N)))

td = time.time() - to
print("Fibonacci Sequence in Array: The time interval is %f s." %td)

#%% Fibonacci with Iterator

to = time.time()
# upLim = 20
class Fibs:
    def __init__(self):
        self.c = 0
        self.a = 0
        self.b = 1
        self.n = upLim
    def __iter__(self):
        return self
    def __next__(self):
        self.i = self.a
        self.a = self.b
        self.b = self.i + self.b
        self.c = self.c + 1
        if self.c > self.n:
            raise StopIteration
        return self.a

fibs = Fibs()
td = time.time() - to
print("Fibonacci with Iterator: The time interval is %f s." %td)

P = np.array(list(fibs))
T = np.arange(1, upLim+1)
print(P)
print(T)

#%%  Plot

upLim = 40
N = np.arange(1,upLim+1,1)
M = np.array(list(map(fib,N)))

plt.rcParams['font.family'] = 'CMU Serif'
plt.rcParams['text.usetex'] = True
fig, ax = plt.subplots(1, 1)
plt.plot(N, M, c="#00B4DC", alpha=0.9)
plt.tick_params(direction='in')

plt.show()